File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/ComboBox.java
package com.spvessel;

import com.spvessel.Common.DefaultsService;
import com.spvessel.Core.*;
import com.spvessel.Decorations.Indents;
import com.spvessel.Decorations.Style;
import com.spvessel.Flags.ItemAlignment;
import com.spvessel.Flags.KeyCode;
import com.spvessel.Flags.MouseButton;
import com.spvessel.Flags.SizePolicy;

import java.awt.Color;
import java.awt.Font;
import java.util.List;

public class ComboBox extends Prototype {
    // Queue<BaseItem> _queue = new Queue<BaseItem>();

    static int count = 0;
    public ButtonCore _selected = new ButtonCore();
    public ButtonCore _dropdown = new ButtonCore();
    public CustomShape _arrow = new CustomShape();
    public ComboBoxDropDown _dropdownarea;
    public EventCommonMethod selectionChanged = new EventCommonMethod();

    /**
     * Constructs a ComboBox
     */
    public ComboBox() {
        setBackground(0, 0, 0, 0);
        setItemName("ComboBox_" + count);
        count++;

        eventKeyPress.add(this::onKeyPress);
        eventMousePress.add((sender, args) -> showDropDownList());

        setStyle(DefaultsService.getDefaultStyle(ComboBox.class));
    }

    void onKeyPress(InterfaceItem sender, KeyArgs args) {
        if (args.key == KeyCode.ENTER && eventMouseClick != null) {
            eventMouseClick.execute(this, new MouseArgs());
        }
    }

    // text init
    /**
     * Text alignment in the ComboBox
     */
    public void setTextAlignment(ItemAlignment... alignment) {
        _selected.setTextAlignment(alignment);
    }

    public void setTextAlignment(List<ItemAlignment> alignment) {
        _selected.setTextAlignment(alignment);
    }

    /**
     * Text margin in the ComboBox
     */
    public void setTextMargin(Indents margin) {
        _selected.setMargin(margin);
    }

    /**
     * Text font parameters in the ComboBox
     */
    public void setFont(Font font) {
        _selected.setFont(font);
    }

    public void setFontSize(int size) {
        _selected.setFontSize(size);
    }

    public void setFontStyle(int style) {
        _selected.setFontStyle(style);
    }

    public void setFontFamily(String font_family) {
        _selected.setFontFamily(font_family);
    }

    public Font getFont() {
        return _selected.getFont();
    }

    /**
     * Set text in the ComboBox
     */
    public void setText(String text) {
        _selected.setText(text);
    }

    public String getText() {
        return _selected.getText();
    }

    /**
     * Text color in the ComboBox
     */
    public void setForeground(Color color) {
        _selected.setForeground(color);
    }

    public void setForeground(int r, int g, int b) {
        _selected.setForeground(r, g, b);
    }

    public void setForeground(int r, int g, int b, int a) {
        _selected.setForeground(r, g, b, a);
    }

    public void setForeground(float r, float g, float b) {
        _selected.setForeground(r, g, b);
    }

    public void setForeground(float r, float g, float b, float a) {
        _selected.setForeground(r, g, b, a);
    }

    public Color getForeground() {
        return _selected.getForeground();
    }

    /**
     * Initialization and adding of all elements in the ComboBox
     */
    @Override
    public void initElements() {
        // adding
        super.addItem(_selected);
        super.addItem(_dropdown);
        _dropdown.addItem(_arrow);

        // dropdownarea
        _dropdownarea = new ComboBoxDropDown(getHandler());
        _dropdownarea.activeButton = MouseButton.BUTTON_LEFT;
        _dropdownarea.setOutsideClickClosable(true);
        _dropdownarea.selectionChanged.add(() -> onSelectionChanged());
    }

    private void showDropDownList() {
        
        // dropdownarea
        _dropdownarea.setPosition(getX(), getY() + getHeight());
        _dropdownarea.setSize(_selected.getWidth(), 100);
        _dropdownarea.setSizePolicy(SizePolicy.FIXED, SizePolicy.FIXED);
        MouseArgs args = new MouseArgs();
        args.button = MouseButton.BUTTON_LEFT;
        _dropdownarea.show(this, args);
    }

    /**
     * Add item to ComboBox list
     */
    @Override
    public void addItem(InterfaceBaseItem item) {
        _dropdownarea.addItem(item);
        // _queue.Enqueue(item);
    }

    /**
     * Remove item from the ComboBox list
     */
    @Override
    public void removeItem(InterfaceBaseItem item) {
        _dropdownarea.removeItem(item);
    }

    /**
     * Current element in the ComboBox by index
     */
    public void setCurrentIndex(int index) {
        _dropdownarea.setCurrentIndex(index);
        _selected.setText(_dropdownarea.getText());
        selectionChanged.execute();
    }

    public int getCurrentIndex() {
        return _dropdownarea.getCurrentIndex();
    }

    private void onSelectionChanged() {
        _selected.setText(_dropdownarea.getText());
        selectionChanged.execute();
    }

    /**
     * Set style of the ComboBox
     */
    @Override
    public void setStyle(Style style) {
        if (style == null)
            return;

        super.setStyle(style);
        setForeground(style.foreground);
        setFont(style.font);

        Style inner_style = style.getInnerStyle("selection");
        if (inner_style != null) {
            _selected.setStyle(inner_style);
        }
        inner_style = style.getInnerStyle("dropdownbutton");
        if (inner_style != null) {
            _dropdown.setStyle(inner_style);
        }
        inner_style = style.getInnerStyle("arrow");
        if (inner_style != null) {
            _arrow.setStyle(inner_style);
        }
    }
}

File: CSharp_docOnly/Items/ComboBox.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using SpaceVIL.Common;
using SpaceVIL.Core;
using SpaceVIL.Decorations;

namespace SpaceVIL
{
    public class ComboBox : Prototype
    {
        static int count = 0;
        public ButtonCore _selected = new ButtonCore();
        public ButtonCore _dropdown = new ButtonCore();
        public CustomShape _arrow;
        public ComboBoxDropDown _dropdownarea;
        public EventCommonMethod SelectionChanged;

        /// <summary>
        /// Constructs a ComboBox
        /// </summary>
        public ComboBox()
        {
            SetBackground(Color.Transparent);
            SetItemName("ComboBox_" + count);
            count++;

            EventKeyPress += OnKeyPress;
            EventMousePress += (sender, args) => ShowDropDownList();

            _arrow = new CustomShape();
            SetStyle(DefaultsService.GetDefaultStyle(typeof(SpaceVIL.ComboBox)));
        }

        void OnKeyPress(object sender, KeyArgs args)
        {
            if (args.Key == KeyCode.Enter)
                EventMouseClick?.Invoke(this, new MouseArgs());
        }

        //text init
        /// <summary>
        /// Text alignment in the ComboBox
        /// </summary>
        public void SetTextAlignment(ItemAlignment alignment)
        {
            _selected.SetTextAlignment(alignment);
        }

        /// <summary>
        /// Text margin in the ComboBox
        /// </summary>
        public void SetTextMargin(Indents margin)
        {
            _selected.SetMargin(margin);
        }

        /// <summary>
        /// Text font parameters in the ComboBox
        /// </summary>
        public void SetFont(Font font)
        {
            _selected.SetFont(font);
        }
        public void SetFontSize(int size)
        {
            _selected.SetFontSize(size);
        }
        public void SetFontStyle(FontStyle style)
        {
            _selected.SetFontStyle(style);
        }
        public void SetFontFamily(FontFamily font_family)
        {
            _selected.SetFontFamily(font_family);
        }
        public Font GetFont()
        {
            return _selected.GetFont();
        }

        /// <summary>
        /// Set text in the ComboBox
        /// </summary>
        public void SetText(String text)
        {
            _selected.SetText(text);
        }
        public String GetText()
        {
            return _selected.GetText();
        }

        /// <summary>
        /// Text color in the ComboBox
        /// </summary>
        public void SetForeground(Color color)
        {
            _selected.SetForeground(color);
        }
        public void SetForeground(int r, int g, int b)
        {
            _selected.SetForeground(r, g, b);
        }
        public void SetForeground(int r, int g, int b, int a)
        {
            _selected.SetForeground(r, g, b, a);
        }
        public void SetForeground(float r, float g, float b)
        {
            _selected.SetForeground(r, g, b);
        }
        public void SetForeground(float r, float g, float b, float a)
        {
            _selected.SetForeground(r, g, b, a);
        }
        public Color GetForeground()
        {
            return _selected.GetForeground();
        }

        /// <summary>
        /// Initialization and adding of all elements in the ComboBox
        /// </summary>
        public override void InitElements()
        {
            //adding
            base.AddItem(_selected);
            base.AddItem(_dropdown);
            _dropdown.AddItem(_arrow);

            //dropdownarea
            _dropdownarea = new ComboBoxDropDown(GetHandler());
            _dropdownarea.SetOutsideClickClosable(true);
            _dropdownarea.SelectionChanged += OnSelectionChanged;
        }

        private void ShowDropDownList()
        {
            //dropdownarea
            _dropdownarea.SetPosition(GetX(), GetY() + GetHeight());
            _dropdownarea.SetSize(_selected.GetWidth(), 100);
            MouseArgs args = new MouseArgs();
            args.Button = MouseButton.ButtonLeft;
            _dropdownarea.Show(this, args);
        }

        /// <summary>
        /// Add item to ComboBox list
        /// </summary>
        public override void AddItem(IBaseItem item)
        {
            _dropdownarea.AddItem(item);
        }

        /// <summary>
        /// Remove item from the ComboBox list
        /// </summary>
        public override void RemoveItem(IBaseItem item)
        {
            _dropdownarea.RemoveItem(item);
        }

        /// <summary>
        /// Current element in the ComboBox by index
        /// </summary>
        public void SetCurrentIndex(int index)
        {
            _dropdownarea.SetCurrentIndex(index);
            _selected.SetText(_dropdownarea.GetText());
            SelectionChanged?.Invoke();
        }
        public int GetCurrentIndex()
        {
            return _dropdownarea.GetCurrentIndex();
        }
        private void OnSelectionChanged()
        {
            _selected.SetText(_dropdownarea.GetText());
            SelectionChanged?.Invoke();
        }

        /// <summary>
        /// Set style of the ComboBox
        /// </summary>
        public override void SetStyle(Style style)
        {
            if (style == null)
                return;

            base.SetStyle(style);
            SetForeground(style.Foreground);
            SetFont(style.Font);

            Style inner_style = style.GetInnerStyle("selection");
            if (inner_style != null)
            {
                _selected.SetStyle(inner_style);
            }
            inner_style = style.GetInnerStyle("dropdownbutton");
            if (inner_style != null)
            {
                _dropdown.SetStyle(inner_style);
            }
            inner_style = style.GetInnerStyle("arrow");
            if (inner_style != null)
            {
                _arrow.SetStyle(inner_style);
            }
        }
    }
}
